README
NAME
    Runops::Recorder - Runops replacement which saves what is being
    performed

SYNOPSIS
      # will save to a rr-<timestamp> directory in the current directory
      perl -MRunops::Recorder <program>

      # will save to a custom directory 
      perl -MRunops::Recorder=my_recording <program>
  
      # and then to view the recording
      rr-viewer <path to recording>

DESCRIPTION
    Runops::Recorder is an alternative runops which saves what it does into
    a file that can later be viewed using the rr-viewer tool.

HOW TO RECORD
    Simply use this module and it'll replace perl's standard runloop with
    its own. By default a recording goes into a directory named
    rr-<date>-_<time>. If you want an alternate name just pass it as the
    first argument to the use (eg -MRunops::Recorder=foo).

    Sometimes perl will optimize away COPs and this may look confusing when
    viewing. If you want to turn of the optimizer pass "-noopt" when using
    this module.

    It's possible to adjust the buffer size which is how much it'll keep in
    memory before flushing it to disk. This is done by passing "-bs=SIZE"
    where SIZE is a number followed by an optional G/M/K/g/m/k to denote the
    multiple. G, M and K are base-2 and g, m, k base-10. So 512K would be
    524288 bytes. If ommited a default of 64K is used. The minimum size is
    128 bytes.

    It is possible to prevent continous store to disk with "-nostore". This
    is ment to be used with "-die" that dumps the buffer to disk when an
    exception occurs.

VIEWING THE RECORDING
    Use the 'rr-viewer' tool. It just takes the path with the recording as
    an argument. Press 'q' to quit or any other key to step to the next
    event. Press 's' to skip any events in the current file until end of
    recording. Press 'a' to toggle wether we should skip whatever is in @INC
    when the recorder what loaded. Press 'h' for help.

    The environment variable RR_AUTORUN tells the viewer to run automaticly.
    The value should be the sleep time until stepping. And yes, it uses
    Time::HiRes so you can give it fracitonal seconds.

    If you set RR_SKIP_INC the autorun will not show @INC files as the 'a'
    option does.

TODO
    Record more things such as changes to variables, opened file descriptors
    etc.

AUTHOR
    Claes Jakobsson, <claesjac@cpan.org>

COPYRIGHT AND LICENSE
    Copyright (C) 2011 by Claes Jakobsson

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either Perl version 5.10.0 or, at
    your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension Runops::Recorder.

0.10  ...
    - Added -ignoreeval which prevents -die to dump buffer when in eval
    - Fixed actual compat with 5.10 and 5.12 claimed.
    - Added at least 5.8.9 compat as requested by Hugo van der Sanden (hopefully
      this means 5.8.8 too)
    
0.09  Sun Nov 13
    - Fixed a warning in the viewer if we haven't seen a TZ yet
    - Added dump() function to dump the buffer to disk
    - Added missing dependencies to Makefile.PL
    
0.08  Sat Nov 12
    - Added an exception viewer which dumps where die was called
    - Timestamp events are now splitted in two - one with seconds and one with
      miliseconds
    - A keyframe is now followed with an switch_file event if we have a file
    - Added -nostore option which prevents continous store to disk. This should 
      be used in combination with -die to only dump snapshots on exceptions
    - Buffer size is configureable via -bs option
    
0.07  Wed Aug 17
    - Keyframes now also record a timestamp so we can search thru our recording
      easier
    - Only the initial keyframe was ever written as I forgot to increment the keyframe
      counter
      
0.06  Wed Aug 17
    - Silence warning 'Use of uninitialized value $opts[0]' when loading without arguments
    - RR_AUTORUN and RR_SKIP_INC are now available as -a <delay> and -s arguments to rr-viewer
    - When done viewing the recording show the last place we were at
    
0.05  Wed Jun 29
    - Improved documentation on how to use this
    - 'die' is now recorded
    - Events are now written to a buffer before being written down to disk
      to reduce the I/O
    - Subroutine names are now fully qualified
    
0.04  Mon Jun 27
    - The format of recordings has changed and is now a directory with a bunch
      of files instead.
    - Added a Runops::Recorder::Reader class that can read recordings and
      callback on different events
    - Now also records when you enter a sub and the name of it
    - A new rr-diff tool that shows two recordings side by side
    
0.03  Sun Jun 19
    - Adjust number of lines shown to fill the screen if possible
    - Added help page 'h'
    - Viewer now support RR_AUTORUN and RR_SKIP_INC environment variables
      which sets how fast it should autorun and if it should skip @INC
      
0.02  Sun Jun 19 11:53 2011
    - Added possibility to skip viewing of a the current file with 's'
    - Viewer adapts to the number of lines your terminal
    - Record when OP_DIE happens
    - Make it possible to skip installed (ie, what's in @INC when we load)
    
0.01  Sat Jun 18 16:07:57 2011
	- original version; created by h2xs 1.23 with options
		-n Runops::Recorder



t/data/example.pl
#!/usr/bin/perl

use strict;
use warnings;

sub foo {
    
}

foo();

1;

pre-5.14-compat.h
/*    pp_ctl.c
 *
 *    Copyright (C) 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000,
 *    2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008 by Larry Wall and others
 *
 *    You may distribute under the terms of either the GNU General Public
 *    License or the Artistic License, as specified in the README file.
 *
 */

#define PERL_ARGS_ASSERT_DOPOPTOSUB_AT	\
	assert(cxstk)

#define dopoptosub(plop)	dopoptosub_at(cxstack, (plop))
#define dopoptosub_at(a,b)	S_dopoptosub_at(aTHX_ a,b)

STATIC I32
S_dopoptosub_at(pTHX_ const PERL_CONTEXT *cxstk, I32 startingblock)
{
    dVAR;
    I32 i;

    PERL_ARGS_ASSERT_DOPOPTOSUB_AT;

    for (i = startingblock; i >= 0; i--) {
	register const PERL_CONTEXT * const cx = &cxstk[i];
	switch (CxTYPE(cx)) {
	default:
	    continue;
	case CXt_EVAL:
	case CXt_SUB:
	case CXt_FORMAT:
	    DEBUG_l( Perl_deb(aTHX_ "(dopoptosub_at(): found sub at cx=%ld)\n", (long)i));
	    return i;
	}
    }
    return i;
}

/*
=for apidoc caller_cx

The XSUB-writer's equivalent of L<caller()|perlfunc/caller>. The
returned C<PERL_CONTEXT> structure can be interrogated to find all the
information returned to Perl by C<caller>. Note that XSUBs don't get a
stack frame, so C<caller_cx(0, NULL)> will return information for the
immediately-surrounding Perl code.

This function skips over the automatic calls to C<&DB::sub> made on the
behalf of the debugger. If the stack frame requested was a sub called by
C<DB::sub>, the return value will be the frame for the call to
C<DB::sub>, since that has the correct line number/etc. for the call
site. If I<dbcxp> is non-C<NULL>, it will be set to a pointer to the
frame for the sub call itself.

=cut
*/

const PERL_CONTEXT *
Perl_caller_cx(pTHX_ I32 count, const PERL_CONTEXT **dbcxp)
{
    register I32 cxix = dopoptosub(cxstack_ix);
    register const PERL_CONTEXT *cx;
    register const PERL_CONTEXT *ccstack = cxstack;
    const PERL_SI *top_si = PL_curstackinfo;

    for (;;) {
	/* we may be in a higher stacklevel, so dig down deeper */
	while (cxix < 0 && top_si->si_type != PERLSI_MAIN) {
	    top_si = top_si->si_prev;
	    ccstack = top_si->si_cxstack;
	    cxix = dopoptosub_at(ccstack, top_si->si_cxix);
	}
	if (cxix < 0)
	    return NULL;
	/* caller() should not report the automatic calls to &DB::sub */
	if (PL_DBsub && GvCV(PL_DBsub) && cxix >= 0 &&
		ccstack[cxix].blk_sub.cv == GvCV(PL_DBsub))
	    count++;
	if (!count--)
	    break;
	cxix = dopoptosub_at(ccstack, cxix - 1);
    }

    cx = &ccstack[cxix];
    if (dbcxp) *dbcxp = cx;

    if (CxTYPE(cx) == CXt_SUB || CxTYPE(cx) == CXt_FORMAT) {
        const I32 dbcxix = dopoptosub_at(ccstack, cxix - 1);
	/* We expect that ccstack[dbcxix] is CXt_SUB, anyway, the
	   field below is defined for any cx. */
	/* caller() should not report the automatic calls to &DB::sub */
	if (PL_DBsub && GvCV(PL_DBsub) && dbcxix >= 0 && ccstack[dbcxix].blk_sub.cv == GvCV(PL_DBsub))
	    cx = &ccstack[dbcxix];
    }

    return cx;
}


lib/Runops/Recorder/Reader.pm
package Runops::Recorder::Reader;

use strict;
use warnings;

use Fcntl qw(SEEK_CUR SEEK_END SEEK_SET);
use File::Spec;

use accessors::ro qw(identifiers_fh data_fh identifiers handler skip_keyframes);

require Exporter;
our @ISA = qw(Exporter);

our @EXPORT = qw();
our @EXPORT_OK = qw(
    KEYFRAME SWITCH_FILE NEXT_STATEMENT DIE ENTER_SUB KEYFRAME_TZ
);

our %EXPORT_TAGS = (
    events => [@EXPORT_OK]
);

use constant {
    KEYFRAME        => 0,
    SWITCH_FILE     => 1,
    NEXT_STATEMENT  => 2,
    DIE             => 3,
    ENTER_SUB       => 4,
    KEYFRAME_TZ     => 5,
    PADSV           => 6,
    LEAVE_SUB       => 7,
    KEYFRAME_TZ_U   => 8,
};

sub new {
    my ($pkg, $path, $opts) = @_;
    
    $opts = {} unless ref $opts eq 'HASH';
    
    my $file = "main.data";
    if (-f $path) {
        (undef, $path, $file) = File::Spec->splitpath($path);
    }
    
    open my $data_fh, "<", File::Spec->catfile($path, $file) 
        or die "Can't read $path/$file because of: $!";
    open my $identifiers_fh, "<", File::Spec->catfile($path, "main.identifiers") 
        or die "Can't read $path/main.identifiers because of: $!";
    
    my $handler;
    if ($opts->{handler}) {
        $handler = _make_class_handler($opts->{handler});
    }
    elsif ($opts->{handlers}) {
        $handler = _make_callback_handler($opts->{handlers});
    }
    
    # This means that read_next will ignore keyframe commands
    my $skip_keyframes = defined $opts->{skip_keyframes} ? $opts->{skip_keyframes} : 1;
    
    my $self = bless { 
        data_fh => $data_fh, 
        identifiers_fh => $identifiers_fh,
        identifiers => {},
        handler => $handler,
        skip_keyframes => $skip_keyframes,
    }, $pkg;

    $self->find_next_keyframe;
    $self->read_identifiers;
    
    return $self;
}

{
    my %CMD_TO_NAME = (
        0 => 'on_keyframe',
        1 => 'on_switch_file',
        2 => 'on_next_statement',
        3 => 'on_die',
        4 => 'on_enter_sub',
        5 => 'on_keyframe_timestamp',
        6 => 'on_padsv',
        7 => 'on_leave_sub',
        8 => 'on_keyframe_timestamp_usec',
    );
    
    my %CMD_DATA_TRANSFORMER = (
        0 => sub {},
        1 => sub { my ($file_no) = unpack("L", $_[0]); return ($file_no, $_[1]->get_identifier($file_no)) },
        2 => sub { my ($line_no) = unpack("L", $_[0]); return ($line_no) },
        3 => sub {},
        4 => sub { my ($identifier_no) = unpack("L", $_[0]); return ($identifier_no, $_[1]->get_identifier($identifier_no)) },
        5 => sub { my ($sec) = unpack("L", $_[0]); return ($sec); },
        6 => sub { my ($identifier_no) = unpack("L", $_[0]); return ($identifier_no, $_[1]->get_identifier($identifier_no)) },
        7 => sub { my ($identifier_no) = unpack("L", $_[0]); return ($identifier_no, $_[1]->get_identifier($identifier_no)) },
        8 => sub { my ($usec) = unpack("L", $_[0]); return ($usec); },
    );
    
    sub _make_class_handler {
        my ($target) = @_;
    
        my $handler = sub {
            my ($cmd, $data, $reader) = @_;
            my $cb = $target->can($CMD_TO_NAME{$cmd});
            $cb->($target, $CMD_DATA_TRANSFORMER{$cmd}->($data, $reader), $reader) if $cb;
            1;
        };
    
        return $handler;
    }
    
    sub _make_callback_handler {
        my ($target) = @_;
        
        my %callbacks = map { 
            my $name = $CMD_TO_NAME{$_};
            ($_, exists $target->{$name} ? $target->{$name} : sub {})
        } keys %CMD_TO_NAME;
        
        my $handler = sub {
            my ($cmd, $data, $reader) = @_;
            $callbacks{$cmd}->($CMD_DATA_TRANSFORMER{$cmd}->($data, $reader), $reader);
        };
    }
    
    sub decode {
        my ($self, $cmd, $data) = @_;
        return $CMD_DATA_TRANSFORMER{$cmd}->($data, $self);
    }
}

sub read_identifiers {
    my $self = shift;
    
    my $identifiers_fh = $self->identifiers_fh;
    
    # Reset that we're not at EOF
    seek $identifiers_fh, 0, SEEK_CUR;
    while (<$identifiers_fh>) {
        chomp;
        my ($id, $name) = split /:/, $_, 2;
        $self->{identifiers}->{$id} = $name;
    }
}

sub get_identifier {
    my ($self, $id) = @_;
    $self->read_identifiers();
    return $self->identifiers->{$id};
}

my @read_extra = (
    0, 
    0, 
    0, 
    0,
    0,
    0,
    0,
    0, 
);

sub read_next {
    my $self = shift;
    
    # Assume the file is synced
    my $buff;
    my $read = read $self->data_fh, $buff, 5;
    if ($read == 5) {
        # decode
        my ($cmd, $data) = unpack("Ca*", $buff);
        # This is a keyframe, return next instead
        return $self->read_next if $cmd == 0 && $self->skip_keyframes;
        if ($read_extra[$cmd]) {
            $self->data_fh->read($buff, $read_extra[$cmd]);
            $data .= $buff;
        }
        $self->handler->($cmd, $data, $self) if $self->handler;
        return ($cmd, $data);
    }
    elsif ($read) {
        seek $self->data_fh, -$read, SEEK_CUR;
    }
    
    return;
}

sub read_all {
    my $self = shift;
    while ($self->read_next()) {
        # nop
    }
}

sub skip_until {
    my ($self, $target_cmd) = @_;
    
    my ($cmd, $data);
    do {
        ($cmd, $data) = $self->read_next();
        return unless defined $cmd;
    }
    until ($cmd == $target_cmd);

    seek $self->data_fh, -5, SEEK_CUR;
}

sub find_next_keyframe {
    my $self = shift;
    
    my $data_fh = $self->data_fh;
    
    my $read_keyframe = 0;

    # TODO: also handle tail mode
    while ($read_keyframe < 5) {
        my $next = getc $data_fh;
        last unless defined $next;
        $read_keyframe = 0, next if ord($next) != 0;
        $read_keyframe++;        
    }
    
    seek $data_fh, -5, SEEK_CUR;

    1;
}

1;
__END__
=pod

=head1 NAME

Runops::Recorder::Reader - A class which can read the recording files

=head1 DESCRIPTION

Instances of this class reads a recording. It can work both as a stream-based 
reader where you ask for the next entry or as a event generator that calls 
your handlers for each type of item it reads.

=head1 SYNOPSIS

  # main script
  use Runops::Recorder::Reader;
  
  my $reader = Runops::Recorder::Reader->read("my-recording", { 
    handler => "MyRecordingHandler",
  });
  
  $reader->read();
  
  # MyRecordingHandler.pm
  package MyRecordingHandler;
  
  sub on_switch_file {
    my ($self, $id, $path) = @_;
    print "Now in file: $path\n";
  }
  
  sub on_next_statement {
    my ($self, $line_no) = @_;
    print "Executing line: $line_no\n";
  }
  
  1;
  
=head1 INTERFACE

=head2 CLASS METHODS

=over 4

=item new ( $path [, \%opts ] )

Creates a new instance of this class. Takes I<$path> which must be a path to a 
recording and an optional hashref with options. Valid options are

=over 4

=item handler

An package name or instance of a class which on which methods will be called when 
events occur. See L</EVENTS>

=item handlers

A hashref with event/callback pairs that are called when events occur. See L</EVENTS>

=item skip_keyframes

An boolean that indicates wheter keyframes should be skipped or not - ie, not 
generate an event or be returned from C<read_next>

=back

=back

=head2 INSTANCE METHODS

=over 4

=item read_next

Reads the next entry in the recording and returns a list with the numeric event and 
its decoded contents. See L</EVENTS>

=item read_all

Reads thru the recording generating events.

=item read_identifiers

Reads the identifiers, files etc that we saw during recording

=item skip_until ( $event )

Reads all events until the next of of type I<$event> occurs.

=item get_identifier ( $id ) 

Returns the identifier with the given I<$id>.

=item find_next_keyframe 

Searches for the next keyframe. This is to be used in the future when one can 
tail recordings being generated.

=item decode ( $event, $data )

Decodes the blob I<$data> according to the rules for the specific event and returns a list of values.

=back

=head1 EVENTS

The following events may occur in the recording. They can be returned by C<read_next> or cause a 
callback/method to be invoked.

The kind of events that can happen. Numeric code, constant and callback name within parenthesis for each item.

=over 4

=item Keyframe (0, C<KEYFRAME>, C<on_keyframe>)

A keyframe is an entry that we can wait for in tailing mode to know where we can start reading. These are 
inserted every 1024 events or so. No data/arguments.

=item Switch file (1, C<SWITCH_FILE>, C<on_switch_file>)

Happens when we execute a statement in another source file than the current one. No arguments.

=item Next statment (2, C<NEXT_STATEMENT>, C<on_next_statement>)

A statement has been executed. Data/argument is C<line number>.

=item Die (3, C<DIE>, C<on_die>)

The program threw an exception using C<die>. No data/arguments.

=item Enter subroutine (4, C<ENTER_SUB>, C<on_enter_sub>)

A subroutine was called. Data/arguments are C<identifier id> and C<identifier>.

=back

=head1 EXPORT

Nothing is exported by default. The tag I<events> export the event constants listed about in L</EVENTS>.

=cut




lib/Runops/Recorder/Viewer.pm
package Runops::Recorder::Viewer;

use strict;
use warnings;

use Runops::Recorder::Reader;

use Scalar::Util qw(weaken);
use Term::Screen;
use Time::HiRes qw(sleep);

my $screen = Term::Screen->new();
$screen->clrscr();

for my $accessor (qw(
    reader current_file_path current_file all_lines
    num_lines skip_files last_line skip_installed
    autorun
    )) { 
    no strict 'refs'; 
    *{$accessor} = sub { $_[0]->{$accessor}; };
}

sub new {
    my ($pkg, $path, $args) = @_;
    
    $args = {} unless ref $args eq 'HASH';
    
    my $self = bless { 
        skip_files => {},
        last_line => 0,
        skip_installed => $ENV{RR_SKIP_INC} || $args->{skip_installed} || 0,
        autorun => $ENV{RR_AUTORUN} || $args->{autorun} || 0,
    }, $pkg;

    my $reader = Runops::Recorder::Reader->new($path, { handler => $self });

    $self->{reader} = $reader;
    
    return $self;
}

sub on_switch_file {
    my ($self, $id, $path) = @_;
    
    close $self->current_file if $self->current_file;
    $self->{current_file_path} = $path;

    if (-e $self->current_file_path) {
        open my $file, "<", $self->current_file_path or die $!;
        $self->{current_file} = $file;
        my @lines = <$file>;
        $self->{all_lines} = \@lines;
        $self->{num_lines} = @lines;

    }
    else {
        $self->{num_lines} = -1;
        $self->{all_lines} = [];
        $self->{current_file} = undef;
    }    
    
    $self->_show_current_file();
}

sub on_keyframe_timestamp {
    my ($self, $tz_sec) = @_;
    
    my ($sec, $min, $hour, $day, $month, $year) = localtime($tz_sec);
    $self->{last_tz} = sprintf("%04d-%02d-%02d %02d:%02d:%02d", $year + 1900, $month + 1, $day, $hour, $min, $sec);
}

sub on_keyframe_timestamp_usec {
    my ($self, $tz_usec) = @_;
    
    $self->{last_tz} .= sprintf ".%6d", $tz_usec;
}

sub _show_current_file {
    my $self = shift;
    
    $screen->clrscr();
    $screen->at(0, 0);

    if ($self->current_file) {
        $screen->bold->puts($self->current_file_path)->normal();
    }
    else {
        $screen->bold->puts("Can't find file\n")->normal();                
    }

    $screen->at(1, 0)->puts("At: " . $self->{last_tz});
}

{
    my $site_libs = join "|", grep { /^\// } @INC;
    my $site_qr = qr{$site_libs};

    sub on_next_statement {
        my ($self, $line_no) = @_;

        $self->{last_line} = $line_no - 1;

        return unless $self->current_file;
        return if $self->skip_files->{$self->current_file_path};
        return if $self->skip_installed && $self->current_file_path =~ $site_qr;

        $self->_show_current_line();
        $self->_process_key();
    }

    sub _show_current_line {
        my $self = shift;
    
        my $line_no = $self->last_line;
        
        my $screen_cols = $screen->cols;
        my $screen_rows = int(($screen->rows - 4) / 2);
    
        my $from = $line_no > $screen_rows ? $line_no - $screen_rows : 0;
        my $to = $line_no + $screen_rows < $self->num_lines - 1 ? $line_no + $screen_rows : $self->num_lines - 1;
    
        # Adjust to fill screen
        $to += ($screen->rows - 4 - ($to - $from)) if ($to - $from) < $screen->rows - 4;

        $screen->at(1, 0)->puts("At: " . $self->{last_tz});    

        $screen->at(2, 0);
        $screen->clreos();
        my $p = length $self->num_lines;
    
        my $i = 0;
        for my $l ($from..$to) {    
            last if $l > $self->num_lines;
            last unless defined $self->all_lines->[$l];
            $screen->at(2 + $i++, 0);
            $screen->reverse if $l == $line_no;
            my $src = sprintf("% ${p}d: %s", $l, substr($self->all_lines->[$l], 0, $screen_cols - ($p + 2)));
            $screen->puts($src);
            $screen->normal if $l == $line_no;
        }        
    }
}

my %KEY_HANDLER = (
    q => sub { shift->done },
    s => sub { 
        my $self = shift; 
        $self->skip_files->{$self->current_file_path} = 1; 
    },
    a => sub { 
        $_[0]->{skip_installed} ^= 1;
        $screen->at(1, 0)->puts("Skip installed is: " . ((qw(OFF ON)[$_[0]->skip_installed])));
        $screen->getch();
    },
    h => \&_show_help,
);

sub _process_key {
    my $self = shift;
    
    if ($self->{autorun}) {
        sleep $self->{autorun};
        return;
    }
    
    my $k = lc $screen->getch();
    $screen->at(1, 0);

    my $handler = $KEY_HANDLER{$k} // sub {};
    $handler->($self);
    
    1;
}

sub _show_help {
    my $self = shift;
    
    $screen->clrscr();
    $screen->at(0, 0)->puts("Help for 'rr-viewer'");
    $screen->at(2, 0)->puts("'a' - Toggle skip files in \@INC at start");
    $screen->at(3, 0)->puts("'s' - Skip the current file");
    $screen->at(5, 0)->puts("'h' - Show this help");
    $screen->at(6, 0)->puts("'q' - Quit prematurely");
    $screen->at(8, 0)->puts("... press the ANY key to continue ...");
    $screen->getch();
    
    $self->_show_current_file();
    $self->_show_current_line();    
}

sub playback {
    my $self = shift;

    $self->reader->read_all();    
    $self->done;
}

sub done {
    my $self = shift;

    $self->_show_current_file();
    $self->_show_current_line();    
    
    $screen->at(0, 0)->clreol();
    $screen->at(0, 0)->puts("Playback completed, press the ANY key to quit...");

  
    $screen->getch();
    
    exit 0;
}

sub view {
    my ($pkg, $path, $args) = @_;

    my $viewer = $pkg->new($path, $args);
    
    $viewer->playback();
}

1;

lib/Runops/Recorder/Viewer/Diff.pm
package Runops::Recorder::Viewer::Diff;

use strict;
use warnings;

use Data::Dumper qw(Dumper);
use Term::ReadKey;

use Runops::Recorder::Reader;

use accessors::ro qw(
    enter_subs
    lines 
    num_readers
    paths 
    readers 
    reader_id 
    sources
);

sub new {
    my ($pkg, @paths) = @_;
    
    my $self = bless {    
    }, $pkg;
    
    my @readers = map { Runops::Recorder::Reader->new($_, { handler => $self }) } @paths;
    $self->{reader_id} = { map { $readers[$_] => $_ } 0..(@readers - 1) };
    $self->{readers} = { map { $_ => $readers[$_] } 0..(@readers - 1) };
    $self->{num_readers} = @paths;
    $self->{sources} = [];
    return $self;
}

sub reset {
    my $self = shift;
    
    $self->{paths} = [];
    $self->{lines} = [];
    $self->{enter_subs} = [];
}

sub on_switch_file {
    my ($self, $id, $path, $reader) = @_;
    
    $self->paths->[$self->reader_id->{$reader}] = $path;
    
    open my $in, "<", $path or die $!;
    my @src = <$in>;
    $self->sources->[$self->reader_id->{$reader}] = \@src;
    close $in;
}

sub on_next_line {
    my ($self, $line, $reader) = @_;
    
    my $data = $self->sources->[$self->reader_id->{$reader}]->[$line - 1];
    chomp $data if $data;

    $self->lines->[$self->reader_id->{$reader}] = [$line, $data];
}

sub on_enter_sub {
    my ($self, $id, $subname, $reader) = @_;
    $self->enter_subs->[$self->reader_id->{$reader}] = $subname;    
}

sub display {
    my $self = shift;
    
    my ($cols) = Term::ReadKey::GetTerminalSize(*STDOUT);
    my $readers = $self->num_readers;
    my $cols_per_reader = int($cols / $readers);

    my $printed;
    if (@{$self->paths}) {
        for (0..($readers - 1)) {
            my $width = $cols_per_reader - 4;
            if ($self->paths->[$_]) {
                $printed += printf "* %-${width}s  ", substr($self->paths->[$_], 0, $cols_per_reader - 2);
            }
            else {
                $printed += print " " x $cols_per_reader;
            }
        }
    }

    if (@{$self->lines}) {
        for (0..($readers - 1)) {
            my $width = $cols_per_reader - 8;
            if ($self->lines->[$_]) {
                my ($line_no, $source) = @{$self->lines->[$_]};
                $source //= "";
                $printed += printf "% 4d: %-${width}s  ", $line_no, substr($source, 0, $width);
            }
            else {
                $printed += print " " x $cols_per_reader;
            }
        }
    }

    if (@{$self->enter_subs}) {
        for (0..($readers - 1)) {
            my $width = $cols_per_reader - 6;
            if ($self->enter_subs->[$_]) {
                printf "sub % -${width}s ", substr($self->enter_subs->[$_], 0, $width) . " {";
            }
            else {
                print " " x $cols_per_reader;
            }
        }
    }    

    print "\n" if $printed;
}

sub run {
    my $self = shift;

    my @readers = values %{$self->readers};
    
    for (;;) {
        $self->reset;

        my $has_more = 0;
        for my $r (@readers) {
            $has_more |= 1 if $r->read_next;
        }
        
        last unless $has_more;
        
        $self->display();
    }
    
    print "Done\n";
}

1;

lib/Runops/Recorder.pm
package Runops::Recorder;

use strict;
use warnings;

use Carp;
use File::Path qw(make_path);
use POSIX qw(strftime);

use constant DEFAULT_BUFFER_SIZE => 64 * 1024;

our $VERSION = '0.09';

require XSLoader;
XSLoader::load('Runops::Recorder', $VERSION);

sub import {
    my ($pkg, @opts) = @_;
    
    my $target_dir;
    $target_dir = $opts[0] if defined $opts[0] && $opts[0] !~ /^-/;
    
    unless ($target_dir) {
        unless ($ENV{RR_TARGET_DIR}) {
            $target_dir = defined $target_dir ? 
                $target_dir :
                strftime("rr-%Y%m%d_%H%M%S", localtime(time));
        }
        else {
            $target_dir = $ENV{RR_TARGET_DIR};
        }
    }
    
    unless (-e $target_dir && -d $target_dir) {
        make_path $target_dir;
    }
    
    set_target_dir($target_dir);

    my $size = _get_buffer_size(\@opts);
    set_buffer_size($size);
    
    # Set options
    my $opts = 0;

    # This option turns on actual writing to disc for the continous store,
    # it's ment to be disabled when -die is in effect
    $opts |= 0x1;
    $opts ^= 0x1 if grep { $_ eq "-nostore" } @opts;
    
    # This option controls wether the buffer should be dumped into a file 
    # when an exception is thrown
    $opts |= 0x2 if grep { $_ eq "-die"} @opts;
    $opts |= 0x4 if grep { $_ eq "-ignoreeval" } @opts;
    
    set_options($opts);
    
    # Maybe disable optimizer
    $^P = 4 if grep { $_ eq "-noopt" } @opts;
    
    init_recorder();
}

my %Size_multiplier = ( 
    G => 1_073_741_824, M => 1_048_576, K => 1024,
    g => 1_000_000_000, m => 1_000_000, k => 1000,
);

sub _get_buffer_size {
    my $opts = shift;
    
    for my $opt (@$opts) {
        if ($opt =~ /^-bs=(\d+(?:\.\d+)?)([GMK])?/i) {
            return $2 ? $1 * $Size_multiplier{$2} : int $1;    
        }
    }
    
    return DEFAULT_BUFFER_SIZE;
}

1;
__END__
=head1 NAME

Runops::Recorder - Runops replacement which saves what is being performed

=head1 SYNOPSIS

  # will save to a rr-<timestamp> directory in the current directory
  perl -MRunops::Recorder <program>

  # will save to a custom directory 
  perl -MRunops::Recorder=my_recording <program>
  
  # and then to view the recording
  rr-viewer <path to recording>
  
=head1 DESCRIPTION

Runops::Recorder is an alternative runops which saves what it does into a file 
that can later be viewed using the rr-viewer tool.

=head1 HOW TO RECORD

Simply use this module and it'll replace perl's standard runloop with its own. By 
default a recording goes into a directory named rr-<date>-_<time>. If you want an 
alternate name just pass it as the first argument to the use (eg -MRunops::Recorder=foo). 

Sometimes perl will optimize away COPs and this may look confusing when viewing. If you 
want to turn of the optimizer pass C<-noopt> when using this module.

It's possible to adjust the buffer size which is how much it'll keep in memory before flushing 
it to disk. This is done by passing C<-bs=SIZE> where SIZE is a number followed by an optional G/M/K/g/m/k 
to denote the multiple. G, M and K are base-2 and g, m, k base-10. So 512K would be 524288 bytes. 
If ommited a default of 64K is used. The minimum size is 128 bytes.

It is possible to prevent continous store to disk with C<-nostore>. This is ment to be used 
with C<-die> that dumps the buffer to disk when an exception occurs.

=head1 VIEWING THE RECORDING

Use the 'rr-viewer' tool. It just takes the path with the recording as an argument. 
Press 'q' to quit or any other key to step to the next event. Press 's' to skip any 
events in the current file until end of recording. Press 'a' to toggle wether we should 
skip whatever is in @INC when the recorder what loaded. Press 'h' for help.

The environment variable RR_AUTORUN tells the viewer to run automaticly. The value 
should be the sleep time until stepping. And yes, it uses Time::HiRes so you can 
give it fracitonal seconds.

If you set RR_SKIP_INC the autorun will not show @INC files as the 'a' option does.

=head1 FUNCTIONS

=over 4

=item dump ( $name )

Dumps the buffer as I<$name> in the recording. Adds .dump to the $name if 
ommited.

=back

=head1 TODO

Record more things such as changes to variables, opened file descriptors etc.

=head1 AUTHOR

Claes Jakobsson, E<lt>claesjac@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Claes Jakobsson

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


=cut
